File: app/src/main/java/com/example/data/embedding/EmbeddingModelManager.kt
package com.example.data.embedding

import android.content.Context
import android.util.Log
import com.example.data.remote.TokenStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Manages the downloadable on-device EmbeddingGemma model.
 *
 * EmbeddingGemma runs through the Google AI Edge RAG library's [GemmaEmbeddingModel] (requires
 * `localagents-rag` ≥ 0.3.0), which needs two files: the LiteRT `.tflite` weights and a SentencePiece tokenizer. We keep them
 * in the app's external files dir (so they survive reinstalls of data, can be inspected, and — handy
 * for development — can be side-loaded with `adb push`). The ~180MB model is *not* bundled in the
 * APK; the user opts in to the download.
 *
 * The source repo (litert-community/embeddinggemma-300m) is gated under the Gemma license, so an
 * unauthenticated fetch may be rejected. [download] accepts an optional bearer token, and any file
 * can also be side-loaded into [modelDir] out of band.
 */
class EmbeddingModelManager(
    private val context: Context,
    private val tokenStore: TokenStore
) {

    sealed interface State {
        /** Model files not present — the on-device path is unavailable. */
        data object Absent : State
        /** Download in progress. [fraction] is 0f..1f across both files; [label] is a human note. */
        data class Downloading(val fraction: Float, val label: String) : State
        /** Both files present — on-device embedding can run. */
        data object Ready : State
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow(initialState())
    val state: StateFlow<State> = _state.asStateFlow()

    private fun initialState(): State {
        if (isReady()) return State.Ready
        val hasModel = modelFile().exists() && modelFile().length() > 0
        val hasTok = tokenizerFile().exists() && tokenizerFile().length() > 0
        if (!hasModel && !hasTok) return State.Absent
        return State.Failed(
            if (hasModel && !isValidModelFile()) "Downloaded weights are not a valid TFLite model — delete and re-download."
            else "Model files are incomplete — delete and re-download."
        )
    }

    fun modelDir(): File {
        val externalDir = context.getExternalFilesDir(null)
        val base = if (externalDir != null) {
            externalDir
        } else {
            Log.w(TAG, "External storage unavailable, using internal storage")
            context.filesDir
        }
        return File(base, "models").apply { mkdirs() }
    }
    fun modelFile(): File = File(modelDir(), MODEL_FILE)
    fun tokenizerFile(): File = File(modelDir(), TOKENIZER_FILE)

    /** True when both the weights and tokenizer are present and non-empty. */
    fun isReady(): Boolean =
        isValidModelFile() && tokenizerFile().let { it.exists() && it.length() > 0 }

    /**
     * Guards against truncated downloads and HTML error pages mistaken for weights: LiteRT/TFLite
     * files carry a `TFL3` magic at bytes 4–7. Without this check [isReady] could go true on a
     * multi-megabyte gated-repo error body, then [GemmaEmbeddingModel] load fails opaquely at embed time.
     */
    fun isValidModelFile(): Boolean {
        val file = modelFile()
        if (!file.exists() || file.length() < MIN_MODEL_BYTES) return false
        return runCatching {
            file.inputStream().use { input ->
                val header = ByteArray(8)
                if (input.read(header) < 8) return false
                header.copyOfRange(4, 8).decodeToString() == "TFL3"
            }
        }.getOrDefault(false)
    }

    /** Re-derive state from disk (e.g. after a side-loaded push). */
    fun refresh() {
        _state.value = initialState()
    }

    // Serializes downloads: a double-tapped DOWNLOAD/RETRY, or a trigger from both the Settings card
    // and the feed sheet, would otherwise run two coroutines writing the same ".part" temp file and
    // both renaming it onto the model path — corrupting the weights on disk. tryLock (not lock) means
    // a concurrent re-trigger is ignored rather than queued behind the in-flight download.
    private val downloadMutex = Mutex()

    private val http: OkHttpClient by lazy {
        // Dedicated client: no body logging (these are large binaries) and a generous read timeout.
        OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(5, TimeUnit.MINUTES)
            .build()
    }

    /**
     * Downloads the tokenizer (small) then the weights (large), reporting combined progress.
     * Safe to call again after a failure — partial files are overwritten. Returns true on success.
     */
    suspend fun download(overrideToken: String? = null): Boolean = withContext(Dispatchers.IO) {
        if (isReady()) {
            _state.value = State.Ready
            return@withContext true
        }
        // Ignore a re-entrant call while a download is already running (see downloadMutex). The
        // in-flight coroutine keeps driving _state, so the UI is unaffected by the ignored tap.
        if (!downloadMutex.tryLock()) {
            Log.d(TAG, "Download already in progress; ignoring re-entrant request")
            return@withContext false
        }
        try {
        // Resolve the HF access token (BYOK): a freshly pasted one wins (and is persisted so it's
        // entered once), then a previously saved token. No key ships in the APK, so null is possible
        // — fine for an ungated mirror; gated repos surface a clear 401/403 hint below.
        //
        // Trim first: tokens pasted from the HF site routinely carry a trailing newline or spaces.
        // Left intact, that newline makes OkHttp throw on the "Bearer …" header (an illegal header
        // value), so the download dies with a confusing error instead of authenticating — and the
        // tainted value would otherwise be persisted and reused on every retry. Trimming on both the
        // fresh and the saved path also self-heals a previously stored tainted token.
        val pasted = overrideToken?.trim()?.takeIf { it.isNotBlank() }
        val authToken = pasted
            ?: tokenStore.getHuggingFaceToken()?.trim()?.takeIf { it.isNotBlank() }
        try {
            // Tokenizer first — it's tiny and lets us fail fast on auth/network before the big file.
            _state.value = State.Downloading(0f, "Fetching tokenizer…")
            fetch(TOKENIZER_URL, tokenizerFile(), authToken) { _ -> }

            fetch(MODEL_URL, modelFile(), authToken) { frac ->
                _state.value = State.Downloading(frac, "Downloading model… ${(frac * 100).toInt()}%")
            }

            if (isReady()) {
                // Persist only a token we've now confirmed works, so a wrong/expired submission can't
                // poison later blank-field retries with a cached bad token.
                if (pasted != null) runCatching { tokenStore.saveHuggingFaceToken(pasted) }
                _state.value = State.Ready
                true
            } else {
                cleanup()
                _state.value = State.Failed(
                    if (!isValidModelFile()) "Downloaded weights are not a valid TFLite model — check your Hugging Face token and Gemma license, then retry."
                    else "Downloaded files were incomplete"
                )
                false
            }
        } catch (e: Exception) {
            Log.e(TAG, "Model download failed: ${e.message}", e)
            cleanup()
            val msg = e.message ?: ""
            // Distinguish the two gated-repo failure modes — a valid token that hasn't accepted the
            // license (403) is the common case and was previously mis-reported as "token required".
            val hint = when {
                msg.contains("403") ->
                    "Access not granted (403). Your token works, but you must accept the Gemma license: open " +
                        "huggingface.co/litert-community/embeddinggemma-300m, click “Agree and access repository”, " +
                        "then retry. (A fine-grained token also needs “Read access to public gated repos”.)"
                msg.contains("401") ->
                    "Hugging Face token invalid or expired (401). Create a new READ token at " +
                        "huggingface.co/settings/tokens and paste it above."
                else -> msg.ifBlank { "Download failed" }
            }
            _state.value = State.Failed(hint)
            false
        }
        } finally {
            downloadMutex.unlock()
        }
    }

    /** Optional hook so [OnDeviceEmbeddingProvider] can drop its in-memory interpreter when files are removed. */
    var onDeleted: (() -> Unit)? = null

    fun delete() {
        cleanup()
        onDeleted?.invoke()
        _state.value = State.Absent
    }

    private fun cleanup() {
        runCatching { modelFile().delete() }
        runCatching { tokenizerFile().delete() }
    }

    private fun fetch(url: String, dest: File, authToken: String?, onProgress: (Float) -> Unit) {
        val reqBuilder = Request.Builder().url(url)
        if (!authToken.isNullOrBlank()) reqBuilder.header("Authorization", "Bearer $authToken")
        http.newCall(reqBuilder.build()).execute().use { response ->
            if (!response.isSuccessful) throw IllegalStateException("HTTP ${response.code} for $url")
            val body = response.body ?: throw IllegalStateException("Empty body for $url")
            val total = body.contentLength().takeIf { it > 0 }
            val tmp = File(dest.parentFile, dest.name + ".part")
            body.byteStream().use { input ->
                tmp.outputStream().use { output ->
                    val buf = ByteArray(1 shl 16)
                    var read: Int
                    var written = 0L
                    while (input.read(buf).also { read = it } != -1) {
                        output.write(buf, 0, read)
                        written += read
                        if (total != null) onProgress(written.toFloat() / total)
                    }
                }
            }
            if (!tmp.renameTo(dest)) {
                tmp.copyTo(dest, overwrite = true); tmp.delete()
            }
        }
    }

    companion object {
        private const val TAG = "EmbeddingModelManager"

        // Local on-disk names (kept stable regardless of the remote filename).
        const val MODEL_FILE = "embeddinggemma_seq256.tflite"
        const val TOKENIZER_FILE = "sentencepiece.model"

        // ~180MB; guards against a truncated/HTML error page being mistaken for the model.
        private const val MIN_MODEL_BYTES = 1_000_000L

        // litert-community/embeddinggemma-300m — gated under the Gemma license.
        // The repo renamed the seq256 weights to a `_mixed-precision` suffix; the old
        // `embeddinggemma-300M_seq256.tflite` path now 404s. The generic (non-vendor) variant is the
        // portable CPU build, which matches our `useGpu = false` inference path.
        private const val REPO = "https://huggingface.co/litert-community/embeddinggemma-300m/resolve/main"
        const val MODEL_URL = "$REPO/embeddinggemma-300M_seq256_mixed-precision.tflite?download=true"
        const val TOKENIZER_URL = "$REPO/sentencepiece.model?download=true"

        /** Approximate total download size, for the UI. */
        const val APPROX_SIZE_LABEL = "~180 MB"
    }
}

File: ios/Curio/Curio/Intelligence/EmbeddingModelManager.swift
import Foundation
import Observation
import os

/// Manages the downloadable on-device EmbeddingGemma model (LiteRT/Core ML weights + SentencePiece
/// tokenizer). Ports `class EmbeddingModelManager` from `data/embedding/EmbeddingModelManager.kt`.
///
/// The ~180 MB model is **not** bundled in the app; the user opts in to the download. The source repo
/// (`litert-community/embeddinggemma-300m`) is gated under the Gemma license, so an unauthenticated
/// fetch may be rejected. `download(overrideToken:)` accepts an optional bearer token; a 401/403
/// surfaces a clear "gated" hint.
///
/// iOS mapping (DESIGN §6 / CONVENTIONS):
/// - Storage moves from Android's external files dir to **Application Support** (`models/`).
/// - State becomes `@Observable` (`ModelState`) so SwiftUI tracks it directly (replaces `StateFlow`).
/// - Download uses `URLSession.bytes(for:)` streaming to a `.part` file, then an **atomic move** into
///   place (replaces the OkHttp byte-stream + `renameTo`).
/// - Token resolution order is preserved (BYOK): a freshly pasted token wins (persisted only once a
///   download it authenticated has succeeded), then a previously saved token, then `nil` — no token
///   ships in the app bundle.
/// - `MIN_MODEL_BYTES` size floor guards against a truncated / HTML error page being mistaken for the
///   model; the tokenizer is fetched **first** so we fail fast on auth/network before the big file.
///
/// `@MainActor` so the `@Observable` state mutations are isolated to the main actor (the heavy byte
/// streaming happens inside the awaited `URLSession.bytes` calls, off the main thread). The disk
/// helpers are `nonisolated`.
@Observable
@MainActor
final class EmbeddingModelManager {

    /// Mirrors the Kotlin `sealed interface State`.
    enum ModelState: Equatable {
        /// Model files not present — the on-device path is unavailable.
        case absent
        /// Download in progress. `fraction` is 0...1 across both files; `label` is a human note.
        case downloading(fraction: Double, label: String)
        /// Both files present — on-device embedding can run.
        case ready
        case failed(String)
    }

    private let tokenStore: TokenStore
    @ObservationIgnored private nonisolated let logger = Logger(subsystem: "com.curio.app", category: "EmbeddingModelManager")

    /// Observable download/availability state. Seeded from disk at construction.
    private(set) var state: ModelState

    /// Serializes downloads: a double-tapped DOWNLOAD/RETRY, or a trigger from both the Settings
    /// card and the feed sheet, would otherwise run two tasks writing the same ".part" temp file
    /// and both renaming it onto the model path — corrupting the weights on disk. The flag is the
    /// `@MainActor` analogue of the Kotlin `Mutex.tryLock()`: a concurrent re-trigger is ignored
    /// rather than queued behind the in-flight download.
    private var isDownloadInFlight = false

    init(tokenStore: TokenStore) {
        self.tokenStore = tokenStore
        self.state = Self.isReadyOnDisk() ? .ready : .absent
    }

    // MARK: - Disk layout (nonisolated — pure FS helpers)

    /// `Application Support/models/` (created on demand). Mirrors Android's `models` subdir.
    nonisolated static func modelDir() -> URL {
        let base = (try? FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )) ?? FileManager.default.temporaryDirectory
        let dir = base.appendingPathComponent("models", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    nonisolated static func modelFileURL() -> URL { modelDir().appendingPathComponent(MODEL_FILE) }
    nonisolated static func tokenizerFileURL() -> URL { modelDir().appendingPathComponent(TOKENIZER_FILE) }

    nonisolated func modelFileURL() -> URL { Self.modelFileURL() }
    nonisolated func tokenizerFileURL() -> URL { Self.tokenizerFileURL() }

    /// True when both the weights and tokenizer are present and non-empty. Port of `isReady()`.
    nonisolated static func isReadyOnDisk() -> Bool {
        let fm = FileManager.default
        let model = modelFileURL()
        let tokenizer = tokenizerFileURL()
        let modelOk = fm.fileExists(atPath: model.path) && fileSize(model) > MIN_MODEL_BYTES
        let tokenizerOk = fm.fileExists(atPath: tokenizer.path) && fileSize(tokenizer) > 0
        return modelOk && tokenizerOk
    }

    /// Instance accessor for `isReady` (used by `EmbeddingAvailability`).
    nonisolated func isReady() -> Bool { Self.isReadyOnDisk() }

    nonisolated static func fileSize(_ url: URL) -> Int64 {
        let attrs = try? FileManager.default.attributesOfItem(atPath: url.path)
        return (attrs?[.size] as? NSNumber)?.int64Value ?? 0
    }

    /// Re-derive state from disk (e.g. after a side-loaded file). Port of `refresh()`.
    func refresh() {
        state = Self.isReadyOnDisk() ? .ready : .absent
    }

    // MARK: - Download

    /// Downloads the tokenizer (small) then the weights (large), reporting combined progress. Safe to
    /// call again after a failure — partial files are overwritten. Returns `true` on success. Port of
    /// `suspend fun download(overrideToken:)`.
    @discardableResult
    func download(overrideToken: String? = nil) async -> Bool {
        if Self.isReadyOnDisk() {
            state = .ready
            return true
        }

        // Ignore a re-entrant call while a download is already running (see `isDownloadInFlight`).
        // The in-flight task keeps driving `state`, so the UI is unaffected by the ignored tap.
        if isDownloadInFlight {
            logger.debug("Download already in progress; ignoring re-entrant request")
            return false
        }
        isDownloadInFlight = true
        defer { isDownloadInFlight = false }

        // Resolve the HF access token (BYOK): a freshly pasted one wins (and is persisted once the
        // download succeeds, so it's entered once), then a previously saved token. No token ships
        // in the app bundle, so `nil` is possible — fine for an ungated mirror; gated repos surface
        // a clear 401/403 hint below.
        //
        // Trim first: tokens pasted from the HF site routinely carry a trailing newline or spaces.
        // Left intact, that whitespace produces an illegal "Bearer …" header value, so the download
        // dies with a confusing error instead of authenticating — and the tainted value would
        // otherwise be persisted and reused on every retry. Trimming on both the fresh and the
        // saved path also self-heals a previously stored tainted token.
        let pasted = overrideToken?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        let savedToken = await tokenStore.getHuggingFaceToken()?
            .trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        let authToken: String? = pasted ?? savedToken

        do {
            // Tokenizer first — tiny, lets us fail fast on auth/network before the big file.
            state = .downloading(fraction: 0, label: "Fetching tokenizer…")
            try await fetch(urlString: Self.TOKENIZER_URL, dest: tokenizerFileURL(), authToken: authToken) { _ in }

            try await fetch(urlString: Self.MODEL_URL, dest: modelFileURL(), authToken: authToken) { [weak self] frac in
                self?.state = .downloading(
                    fraction: frac,
                    label: "Downloading model… \(Int(frac * 100))%"
                )
            }

            if Self.isReadyOnDisk() {
                // Persist only a token we've now confirmed works, so a wrong/expired submission
                // can't poison later blank-field retries with a cached bad token.
                if let pasted { await tokenStore.saveHuggingFaceToken(pasted) }
                state = .ready
                return true
            } else {
                cleanup()
                state = .failed("Downloaded files were incomplete")
                return false
            }
        } catch is CancellationError {
            cleanup()
            state = .failed("Download failed")
            return false
        } catch {
            logger.error("Model download failed: \(error.localizedDescription, privacy: .public)")
            cleanup()
            let message = (error as? DownloadError)?.message ?? error.localizedDescription
            let hint: String
            if message.contains("401") || message.contains("403") {
                hint = "Model is gated — a Hugging Face access token is required."
            } else {
                hint = message.isEmpty ? "Download failed" : message
            }
            state = .failed(hint)
            return false
        }
    }

    /// Deletes the model files and resets state. Port of `delete()`.
    func delete() {
        cleanup()
        state = .absent
    }

    private func cleanup() {
        try? FileManager.default.removeItem(at: modelFileURL())
        try? FileManager.default.removeItem(at: tokenizerFileURL())
    }

    /// Streams `urlString` to a `.part` file reporting progress, then atomically moves it to `dest`.
    /// Mirrors the Kotlin `fetch(url, dest, authToken, onProgress)`: a non-success status throws an
    /// error whose message embeds the HTTP code (so the 401/403 hint fires), an empty body throws,
    /// and the temp file is renamed (with copy-fallback) into place.
    @MainActor
    private func fetch(
        urlString: String,
        dest: URL,
        authToken: String?,
        onProgress: @escaping @MainActor (Double) -> Void
    ) async throws {
        guard let url = URL(string: urlString) else {
            throw DownloadError("Invalid URL")
        }
        var request = URLRequest(url: url)
        request.timeoutInterval = 300 // generous read timeout for large binaries (5 min in Android)
        if let authToken, !authToken.isBlank {
            request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        }

        // A dedicated session: long resource timeout, no body logging. URLSession is retained by
        // the system until invalidated — without finishTasksAndInvalidate() every download
        // attempt (including each retry) leaked one session for the rest of the process.
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 600
        let session = URLSession(configuration: config)
        defer { session.finishTasksAndInvalidate() }

        let (bytes, response) = try await session.bytes(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw DownloadError("No HTTP response for \(urlString)")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw DownloadError("HTTP \(http.statusCode) for \(urlString)")
        }
        let total = http.expectedContentLength // -1 when unknown
        let knownTotal = total > 0 ? Double(total) : nil

        let tmp = dest.appendingPathExtension("part")
        try? FileManager.default.removeItem(at: tmp)
        FileManager.default.createFile(atPath: tmp.path, contents: nil)
        guard let handle = try? FileHandle(forWritingTo: tmp) else {
            throw DownloadError("Cannot open temp file for \(urlString)")
        }
        defer { try? handle.close() }

        var buffer = Data()
        buffer.reserveCapacity(1 << 16)
        var written: Int64 = 0

        for try await byte in bytes {
            buffer.append(byte)
            if buffer.count >= (1 << 16) {
                try handle.write(contentsOf: buffer)
                written += Int64(buffer.count)
                buffer.removeAll(keepingCapacity: true)
                if let knownTotal { onProgress(Double(written) / knownTotal) }
            }
        }
        if !buffer.isEmpty {
            try handle.write(contentsOf: buffer)
            written += Int64(buffer.count)
            if let knownTotal { onProgress(Double(written) / knownTotal) }
        }
        try handle.close()

        // Empty body guard (Kotlin `body ?: throw … "Empty body"`).
        if written == 0 {
            try? FileManager.default.removeItem(at: tmp)
            throw DownloadError("Empty body for \(urlString)")
        }

        // Atomic move into place (Kotlin `renameTo`, with copy+delete fallback).
        try? FileManager.default.removeItem(at: dest)
        do {
            try FileManager.default.moveItem(at: tmp, to: dest)
        } catch {
            try FileManager.default.copyItem(at: tmp, to: dest)
            try? FileManager.default.removeItem(at: tmp)
        }
    }

    /// Internal download error carrying the message used for the gated-hint substring check.
    private struct DownloadError: Error {
        let message: String
        init(_ message: String) { self.message = message }
    }

    // MARK: - Constants (mirror the Kotlin companion object)

    // `nonisolated` so the pure FS helpers above (also `nonisolated`) can read them; they are
    // immutable `Sendable` constants regardless of the type's `@MainActor` isolation.

    /// Local on-disk names (stable regardless of the remote filename).
    nonisolated static let MODEL_FILE = "embeddinggemma_seq256.tflite"
    nonisolated static let TOKENIZER_FILE = "sentencepiece.model"

    /// ~180 MB; guards against a truncated / HTML error page being mistaken for the model.
    nonisolated static let MIN_MODEL_BYTES: Int64 = 1_000_000

    /// `litert-community/embeddinggemma-300m` — gated under the Gemma license.
    /// The repo renamed the seq256 weights to a `_mixed-precision` suffix; the old
    /// `embeddinggemma-300M_seq256.tflite` path now 404s (mirrors Android `EmbeddingModelManager`).
    nonisolated private static let REPO = "https://huggingface.co/litert-community/embeddinggemma-300m/resolve/main"
    nonisolated static let MODEL_URL = "\(REPO)/embeddinggemma-300M_seq256_mixed-precision.tflite?download=true"
    nonisolated static let TOKENIZER_URL = "\(REPO)/sentencepiece.model?download=true"

    /// Approximate total download size, for the UI.
    nonisolated static let APPROX_SIZE_LABEL = "~180 MB"
}

// MARK: - Blank helper (Kotlin `takeIf { it.isNotEmpty() }`)

private extension String {
    /// `nil` when empty (e.g. a whitespace-only token collapsed by trimming), else the value.
    var nilIfEmpty: String? { isEmpty ? nil : self }
}
